Failed to request validator duties

Issue reported by `@canbo` on Teku Discord.

Context: 10 validators were loaded, but they are not yet activated. I suspect we need to add a check in `ValidatorApiHandler.getDuties()` that the validator is active, or somewhere further down the pipeline may be better.

Version: `teku/v0.12.1-dev-deb61dfc/linux-aarch_64/-ubuntu-openjdk64bitservervm-java-11`

```
18:08:45.867 INFO  - Slot Event  *** Slot: 73093, Block: 572f8b..7a8d, Epoch: 2284, Finalized checkpoint: 2282, Finalized root: 80d0f8..48f8, Peers: 20
18:08:49.051 ERROR - Failed to request validator duties for epoch 2283. Retrying after delay.
java.util.concurrent.CompletionException: java.util.NoSuchElementException: No value present
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:683) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658) ~[?:?]
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2094) ~[?:?]
	at tech.pegasys.teku.util.async.SafeFuture.thenApply(SafeFuture.java:268) ~[teku-util-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.getDuties(ValidatorApiHandler.java:125) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at jdk.internal.reflect.GeneratedMethodAccessor41.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at tech.pegasys.teku.events.DirectEventDeliverer.executeMethod(DirectEventDeliverer.java:70) ~[teku-events-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.events.DirectEventDeliverer.deliverToWithResponse(DirectEventDeliverer.java:63) ~[teku-events-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.events.AsyncEventDeliverer.lambda$deliverToWithResponse$1(AsyncEventDeliverer.java:70) ~[teku-events-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.events.AsyncEventDeliverer$QueueReader.deliverNextEvent(AsyncEventDeliverer.java:111) [teku-events-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.events.AsyncEventDeliverer$QueueReader.run(AsyncEventDeliverer.java:103) [teku-events-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.util.NoSuchElementException: No value present
	at java.util.Optional.orElseThrow(Optional.java:382) ~[?:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.createValidatorDuties(ValidatorApiHandler.java:302) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$getDutiesForValidator$9(ValidatorApiHandler.java:288) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at java.util.Optional.map(Optional.java:265) ~[?:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.getDutiesForValidator(ValidatorApiHandler.java:287) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$getValidatorDutiesFromState$8(ValidatorApiHandler.java:277) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.getValidatorDutiesFromState(ValidatorApiHandler.java:278) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$getDuties$1(ValidatorApiHandler.java:129) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at java.util.Optional.map(Optional.java:265) ~[?:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$getDuties$2(ValidatorApiHandler.java:129) ~[teku-validator-coordinator-0.12.1-SNAPSHOT.jar:0.12.1-dev-deb61dfc]
	at java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:680) ~[?:?]
	... 15 more
```

--- a/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java
+++ b/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java
@@ -51,7 +51,6 @@ import tech.pegasys.teku.datastructures.operations.Attestation;
 import tech.pegasys.teku.datastructures.operations.AttestationData;
 import tech.pegasys.teku.datastructures.operations.SignedAggregateAndProof;
 import tech.pegasys.teku.datastructures.state.BeaconState;
-import tech.pegasys.teku.datastructures.state.CommitteeAssignment;
 import tech.pegasys.teku.datastructures.state.ForkInfo;
 import tech.pegasys.teku.datastructures.util.AttestationUtil;
 import tech.pegasys.teku.datastructures.util.CommitteeUtil;
@@ -297,17 +296,18 @@ public class ValidatorApiHandler implements ValidatorApiChannel {
       final Integer validatorIndex) {
     final List<UnsignedLong> proposerSlots =
         proposalSlotsByValidatorIndex.getOrDefault(validatorIndex, emptyList());
-    final CommitteeAssignment committeeAssignment =
-        CommitteeAssignmentUtil.get_committee_assignment(state, epoch, validatorIndex)
-            .orElseThrow();
-    return ValidatorDuties.withDuties(
-        key,
-        validatorIndex,
-        Math.toIntExact(committeeAssignment.getCommitteeIndex().longValue()),
-        committeeAssignment.getCommittee().indexOf(validatorIndex),
-        getAggregatorModulo(committeeAssignment.getCommittee().size()),
-        proposerSlots,
-        committeeAssignment.getSlot());
+    return CommitteeAssignmentUtil.get_committee_assignment(state, epoch, validatorIndex)
+        .map(
+            committeeAssignment ->
+                ValidatorDuties.withDuties(
+                    key,
+                    validatorIndex,
+                    Math.toIntExact(committeeAssignment.getCommitteeIndex().longValue()),
+                    committeeAssignment.getCommittee().indexOf(validatorIndex),
+                    getAggregatorModulo(committeeAssignment.getCommittee().size()),
+                    proposerSlots,
+                    committeeAssignment.getSlot()))
+        .orElseGet(() -> ValidatorDuties.noDuties(key));
   }
 
   private Map<Integer, List<UnsignedLong>> getBeaconProposalSlotsByValidatorIndex(

--- a/validator/coordinator/src/test/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandlerTest.java
+++ b/validator/coordinator/src/test/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandlerTest.java
@@ -22,6 +22,7 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 import static tech.pegasys.teku.datastructures.util.AttestationProcessingResult.SUCCESSFUL;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.compute_epoch_at_slot;
 import static tech.pegasys.teku.util.async.SafeFuture.completedFuture;
 
 import com.google.common.eventbus.EventBus;
@@ -133,6 +134,19 @@ class ValidatorApiHandlerTest {
         validatorApiHandler.getDuties(EPOCH, List.of(unknownPublicKey));
     final Optional<List<ValidatorDuties>> duties = assertCompletedSuccessfully(result);
     assertThat(duties.get()).containsExactly(ValidatorDuties.noDuties(unknownPublicKey));
+  }
+
+  @Test
+  public void getDuties_shouldReturnDutiesForInactiveValidator() {
+    BeaconState state = createStateWithMixOfActiveAndInactiveValidators();
+    when(chainDataClient.getLatestStateAtSlot(PREVIOUS_EPOCH_START_SLOT))
+        .thenReturn(completedFuture(Optional.of(state)));
+
+    final BLSPublicKey publicKey = state.getValidators().get(0).getPubkey();
+    final SafeFuture<Optional<List<ValidatorDuties>>> result =
+        validatorApiHandler.getDuties(EPOCH, List.of(publicKey));
+    final Optional<List<ValidatorDuties>> duties = assertCompletedSuccessfully(result);
+    assertThat(duties.get()).containsExactly(ValidatorDuties.noDuties(publicKey));
   }
 
   @Test
@@ -390,6 +404,32 @@ class ValidatorApiHandlerTest {
                         validator
                             .withActivation_eligibility_epoch(ZERO)
                             .withActivation_epoch(ZERO)
+                            .withExit_epoch(Constants.FAR_FUTURE_EPOCH)
+                            .withWithdrawable_epoch(Constants.FAR_FUTURE_EPOCH));
+              }
+            });
+  }
+
+  private BeaconState createStateWithMixOfActiveAndInactiveValidators() {
+    return createStateWithMixOfActiveAndInactiveValidators(PREVIOUS_EPOCH_START_SLOT);
+  }
+
+  private BeaconState createStateWithMixOfActiveAndInactiveValidators(final UnsignedLong slot) {
+    final UnsignedLong futureEpoch = compute_epoch_at_slot(slot).plus(UnsignedLong.valueOf(10));
+    return dataStructureUtil
+        .randomBeaconState(32)
+        .updated(
+            state -> {
+              state.setSlot(slot);
+              final SSZMutableList<Validator> validators = state.getValidators();
+              for (int i = 0; i < validators.size(); i++) {
+                final UnsignedLong activationEpoch = i % 2 == 0 ? futureEpoch : ZERO;
+                validators.update(
+                    i,
+                    validator ->
+                        validator
+                            .withActivation_eligibility_epoch(activationEpoch)
+                            .withActivation_epoch(activationEpoch)
                             .withExit_epoch(Constants.FAR_FUTURE_EPOCH)
                             .withWithdrawable_epoch(Constants.FAR_FUTURE_EPOCH));
               }

